documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+
+Matricies **must** fulfill the following requirements:
+- be a struct
+- Stored in row major format
+- F matricies work with F vectors, and I Matricies work with I vectors
+- Both column and row vectors and individual values accessible via properties
+- An indexer that takes row and column indicies and outputs the value
+- Add, subtract, and multiply operators defined with Matricies of the same size
+- Multiply operators defined with compatible matricies, if the output matrix type already exists (AxB * BxC = AxC)
+- Multiply operator defined with the appropriate Vector type (vector4F * Matrix4x4F is valid)
+  - **optionally**, include multiplication operators for Vector types smaller than the appropriate type, assuming 1 for the missing dimensions (vector2F * Matrix4x4F -> (Vector2F.x, Vector2F.y, 1, 1) * Matrix4x4F)
+- Negate Operator defined
+- Fulfills `IEquatable<T>` where `T` is the same matrix class and the System.Numerics version of the matrix if available
+- Implicit conversion to and from the System.Numerics matrix type, if available
+- Invert function for square matricies
+- GetDeterminant function for square matricies and Matrix3x2
+- Transpose function
+- static lerp function
+- static identity property
+- For Matrix3x2, Matrix3x3, Matrix4x3, and Matrix4x4 include the following static functions
+  - CreateBillboard
+  - CreateRotation
+    - 3x3, 4x3, and 4x4 Matricies get X, Y, and Z variants for this function instead
+  - CreateTranslation
+  - CreateScale
+  - Decompose (separate out any transformations)
+- For Matrix3x2 include a CreateSkew static function
+- For Matrix3x3, Matrix4x3, and Matrix4x4 include the following static functions
+  - CreateFromAxisAngle
+  - CreateFromQuaternion
+  - Transform
+    - from a Quaternion
+  - CreateFromYawPitchRoll
+- For Matrix4x3 and Matrix4x4 include the following static functions
+  - CreateConstrainedBillboard
+  - CreateLookAt
+  - CreateOrthographic
+  - CreateOrthographicOffCenter
+  - CreatePerspective
+  - CreatePerspectiveFieldOfView
+  - CreatePerspectiveOffCenter
+  - CreateReflection
+  - CreateWorld
+
+# Quaternion
+
+Quaternion is required only to fulfill the following api:
+
+```cs
+[StructLayout(LayoutKind.Sequential), DataContract, Serializable]
+public readonly struct Quaternion<T> : IEquatable<Quaternion<T>>
+    where T : INumberBase<T>, ITrigonometricFunctions<T>, IRootFunctions<T>
+{
+    /// <summary>The X value of the vector component of the quaternion.</summary>
+    [DataMember]
+    public readonly T X { get; }
+
+    /// <summary>The Y value of the vector component of the quaternion.</summary>
+    [DataMember]
+    public readonly T Y { get; }
+
+    /// <summary>The Z value of the vector component of the quaternion.</summary>
+    [DataMember]
+    public readonly T Z { get; }
+
+    /// <summary>The rotation component of the quaternion.</summary>
+    [DataMember]
+    public readonly T W { get; }
+    
```

Comment:
A number of libraries also expose `Vector3 Axis` (X, Y, Z) and `float Angle` (`2 * Acos(W)`) as "properties"

Many even parity most of the Vector4 surface area, including `Length`